src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -0,0 +1,246 @@
+        ConsumerConfig.EnableAutoCommit = enableAutoCommit;
+        ConsumerConfig.EnableAutoOffsetStore = enableAutoOffsetStore;
+        Consumer = new ConsumerBuilder<Null, byte[]>(ConsumerConfig)
+            .SetKeyDeserializer(Deserializers.Null)
+            .SetValueDeserializer(Deserializers.ByteArray)
+            .SetLogHandler(OnLog)
+            .SetErrorHandler(OnError)
+            .Build();
+
+        Consumer.Subscribe(Topic);
+
+        if (TransportBindingElement.ErrorHandlingStrategy == KafkaErrorHandlingStrategy.DeadLetterQueue)
+        {
+            ProducerConfig producerConfig = new(ConsumerConfig);
+            producerConfig.Acks = Acks.All;
+            Producer = new ProducerBuilder<Null, byte[]>(producerConfig)
+                .SetKeySerializer(Serializers.Null)
+                .SetValueSerializer(Serializers.ByteArray)
+                .Build();
+        }
+
+        Task.Run(async () =>
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                try
+                {
+                    var consumeResult = Consumer.Consume(ct);
+                    if (ConsumerConfig.EnablePartitionEof == true && consumeResult.IsPartitionEOF)
+                    {
+                        continue;
+                    }
+
+                    _logger.LogInformation("Received message from kafka at {topicPartitionOffset}", consumeResult.TopicPartitionOffset);
+                    if (_kafkaDeliverySemantics == KafkaDeliverySemantics.AtMostOnce)
+                    {
+                        Consumer.Commit(consumeResult);
+                    }
+
+                    await OnConsumeMessage(consumeResult, queueTransportContext);
+
+                }
+                catch (OperationCanceledException)
+                {
+                    break;
+                }
+                catch (ConsumeException e)
+                {
+                    if (e.Error.IsFatal)
+                    {
+                        _logger.LogCritical("Exit consume loop {code} {error}", e.Error.Code, e.Error.Reason);
+                        break;
+                    }
+
+                    _logger.LogError(e, "Consume error {code} {error}", e.Error.Code, e.Error.Reason);
+                }
+                catch (Exception e)
+                {
+                    _logger.LogCritical(e, "Unexpected error");
+                    break;
+                }
+            }
+            _mres.Set();
+        }, ct);
+        return Task.CompletedTask;
+    }
+
+    public override async Task StopPumpAsync(CancellationToken token)
+    {
+        _cts.Cancel();
+        await _mres.WaitAsync(CancellationToken.None);
+        _cts.Dispose();
+        _receiveContextCountdownEvent.Signal();
+        _receiveContextCountdownEvent.Wait(CancellationToken.None);
+        _receiveContextCountdownEvent.Dispose();
```

Comment:
You should pass in the CancellationToken passed to StopPumpAsync otherwise you can block indefinitely. This will need a try/catch to catch OperationCanceledException, If the token gets cancelled before you are done, you need to do some kind of unclean shutdown. This means deciding what to do with in flight requests. Do you potentially put them in a DLQ? Do you keep track of in-flight requests and inform Kafka that they weren't completed, then if the executing thread does complete, the OnCompleteAsync/OnAbandonAsync know things were shut down and are a no-op? That's the approach used in multiple places in WCF.

Reply:
@mconnew I wired the KafkaBinding.CloseTimeout to a CancellationTokenSource that I'm using to gracefully shutdown (the DLQ producer have to be flushed and the consumer have to be closed properly to exit the consumer group so the broker know how to distribute the remaining offsets).
The AsyncManualResetEvent _mres is responsible to ensure we properly stop the consumption and the CountdownEvent _receiveContextCountdownEvent is responsible to wait for already fetched messages and apply the appropriate handling.